spec/indexers/hyrax/valkyrie_work_indexer_spec.rb

```diff
@@ -9,16 +9,23 @@
   it_behaves_like 'a Work indexer'
 
   context 'when extending with basic metadata' do
+    before do
+      module Hyrax::Test
+        module Basic
+          class Work < Hyrax::Work
+            include Hyrax::Schema(:basic_metadata)
+          end
+        end
+      end
+    end
+    after { Hyrax::Test.send(:remove_const, :Basic) }
+
     let(:indexer_class) do
       Class.new(described_class) do
         include Hyrax::Indexer(:basic_metadata)
       end
     end
-    let(:resource) do
-      Class.new(Hyrax::Work) do
-        include Hyrax::Schema(:basic_metadata)
-      end.new
-    end
+    let(:resource) { Hyrax::Test::Basic::Work.new }
```

Comment:
This change is required to avoid the following error when attempting to extract the human readable name from the class name.

```
ArgumentError:
       Class name cannot be blank. You need to supply a name argument when anonymous class given
```

Reply:
what's the rest of the stack trace.

i'd rather make the new code resilient to the current test cases than change the expectations on the collaborator objects, all else being equal

Reply:
Looking at the code, I suspect mutation may be a challenging path.  Below are the two relevant chunks of code.

First, the `ValkyrieCreateStrategy#result` method and then the `after(:create)` callback which the `#result` method calls via `evaluation.notify(:after_create, result)`.  There are four relevant after create blocks (`:hyrax_work`, `:hyrax_resource`, `:hyrax_collection`, `:hyrax_file_set`).

Were we using ActiveRecord::Base for the `after(:create)`, within the `after(:create)` we could call `work.reload` and mutate the work easily.  For this to work, we would need to update each of the four `after(:create)` callbacks to overwrite the attributes of the `work` object.

The `result` object in the `ValkyrieCreateStrategy#result` and the `work` objected yield in the `after(:create)` callback have the same `#object_id`.  The `evaluation.object` has a different `object_id`.

Which I wonder if it's more appropriate to mutate the `evaluation.object` instead?  That would mean updating the `persister.save` method.


```ruby
# ./lib/hyrax/specs/shared_specs/factories/strategies/valkyrie_resource.rb
class ValkyrieCreateStrategy
  def result(evaluation)
    evaluation.notify(:after_build, evaluation.object)
    evaluation.notify(:before_create, evaluation.object)

    result = persister.save(resource: evaluation.object)

    evaluation.notify(:after_create, result)
    # Before the change, instead of `query_service.find_by(id: result.id)` we had `result`
    query_service.find_by(id: result.id)
  end
  # ...
end

# ./spec/factories/hyrax_work.rb
FactoryBot.define do
  factory :hyrax_work, class: 'Hyrax::Test::SimpleWork' do
  # ..
  after(:create) do |work, evaluator|
      if evaluator.visibility_setting
        Hyrax::VisibilityWriter
          .new(resource: work)
          .assign_access_for(visibility: evaluator.visibility_setting)
      end

      work.permission_manager.edit_groups = evaluator.edit_groups
      work.permission_manager.edit_users  = evaluator.edit_users
      work.permission_manager.read_users  = evaluator.read_users

      work.permission_manager.acl.save
    end
  # ...
end
```